rfcs/0039-snark-keys-management.md

```diff
@@ -0,0 +1,208 @@
+## Summary
+[summary]: #summary
+
+This RFC proposes some strategies to make our snark key management more
+flexible, and to allow us to be more explicit about which keys to use and when
+to generate them.
+
+## Motivation
+[motivation]: #motivation
+
+The current key management strategy is brittle and hard to understand or
+interact with.
+* Every change that affects the snark keys requires a recompile of the majority
+  of our libraries, and it is not possible to specify changes to the constraint
+  system and the corresponding proving/verification keypairs at
+  runtime.
+* Snark keys are generated as part of the build process for some of the
+  compile-time configurations, which slows down compiles and obfuscates the
+  particular keys that any build is using, as well as preventing us from using
+  a single, unified binary for our different configurations.
+* Every set of keys that have been generated for a CI job are stored in the S3
+  keys bucket, with no easy way to identify them.
+* Keys in the S3 bucket are always preferred to locally-generated keys during
+  the build process.
+  - During local builds from developers' machines, internet connection problems
+    may result in either partially-downloaded keyfiles or long download waits
+    while the keys are fetched from S3.
+* It is difficult for the infrastructure team to find and package the correct
+  keys for a particular binaries when deploying.
+* The caching behavior for keys is implicit and entangled with the pickles
+  proof system interface, which makes it hard for the team to inspect or adjust
+  the behavior.
+
+The goal of this RFC is to resolve these issues. At a high-level, the proposed
+steps to do this are:
+* Make keys identifiable from the file contents.
+* Add a tool for generating keys outside of the build process.
+* Allow new keys to be specified explicitly.
+* Make key generation explicit, removing it from the build process.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+### Make keys identifiable from the file contents
+
+In order to confirm that loaded keys are compatible with a given configuration,
+we need to have some representation of the configuration embedded with the keys
+files. We propose that every key (and URS) file contains a header in minified
+JSON format before the actual contents, specifying
+* the constraint constants, as defined in the runtime configuration
+* the SHA commit identifiers for the current commit in the coda, marlin and
+  zexe repos used when generating the keys
+* the length of the binary data following the header, so that we can validate
+  that the file was successfully downloaded in full, where applicable
+* the date when the file's contents were created
```

Comment:
If we don't include generation date, this (in theory) should be reproducible and I think that's worth it. We're going to want easy reproducibility for the more security-conscious folks and it makes it easy for us to go back to old commits to recover data if needed.

Since we have the commit sha, this can act as a timestamp

Reply:
In the medium term, it would be good to have a `ls`-style tool that lets us list all of the cached keys in different ways, and especially to have `list-keys --sort=date` so that we could delete out-of-date keys more easily and reliably.

I've updated the rfc to use the date from the current commit, so we have reproducibility *and* some notion of time, and mentioned these reasons as sub-bullets.